### ISSUE TITLE
Input::Abstract - with multiple inheritance allowed

### ISSUE THREAD

**Issue by jbrezmorf** [flow123d/flow123d#347]
Prerequisite: #337, #348 

- [x] Find all AbstractRecords still declaring some keys. Move them to childs or introduce auxiliary record to copy keys from.
- [ ] Understand reason why we need "full_type_name" which in turn disallows extension of current implementation to allow multiple inheritance.
- [ ] Try to fix the blocker so that we can allow it.
- [x] Convert AbstratRecord into Abstract, do not derive from Record.
- [ ] Check that we get

**Comment by dflanderka** [comment]
Potential trouble in set TYPE selection in Record::finish():

TYPE can not have type of child selection of the parent abstract record since there can be more of them. On the other hand we want to store the actual type in form of index from the selection of particular parent. To make it compatible with the type of the TYPE key, we can declare this key as Interger in the Record, and change implementation in the JSONTo Storage so that the TYPE key is processed in AbstractRecord and skipped in Record. But there are much more places to check, so we postpone all this.

**Comment by jbrezmorf** [comment]
This change also resolve one comment from the pull request of generic types in Record::finish():
OK, setting  type  of the key TYPE to NULL in derive_from is error prone and needs checks like this one. What about to set it to an empty selection. Then we still can distinguish completed TYPE keys if it is necessary but make it litle bit more safe. This change however needs careful check of all places dealing with the type of TYPE.
We can postpone this change since we want to change Selection to Integer anyway.

**Comment by dflanderka** [comment]
[] Change format of key 'implements' in JSON machine output of Record. It must allow to printout any number of ancestors.

**Comment by jbrezmorf** [comment]
key 'implements' is already an array so it is straight forward
